modules/gapi/src/streaming/vpl/vpl_dx11_accel.cpp

```diff
@@ -0,0 +1,171 @@
+            &context // Returns the device immediate context.
+            );
+    if(FAILED(err))
+    {
+        throw std::logic_error("Cannot create D3D11CreateDevice, error: " + std::to_string(HRESULT_CODE(err)));
+    }
+
+    // oneVPL recommendation
+    ID3D11DeviceContext     *pD11Context;
+    ID3D11Multithread       *pD11Multithread;
+    hw_handle->GetImmediateContext(&pD11Context);
+    pD11Context->QueryInterface(IID_PPV_ARGS(&pD11Multithread));
+    pD11Multithread->SetMultithreadProtected(true);
+ 
+    mfxStatus sts = MFXVideoCORE_SetHandle(session, MFX_HANDLE_D3D11_DEVICE, (mfxHDL) hw_handle);
+    if (sts != MFX_ERR_NONE)
+    {
+        throw std::logic_error("Cannot create VPLDX11AccelerationPolicy, MFXVideoCORE_SetHandle error: " +
+                               mfxstatus_to_string(sts));
+    }
+#else
+    mfxStatus sts;
+#endif
+    sts = MFXVideoCORE_GetHandle(session, MFX_HANDLE_D3D11_DEVICE, reinterpret_cast<mfxHDL*>(&hw_handle));
+    if (sts != MFX_ERR_NONE)
+    {
+        throw std::logic_error("Cannot create VPLDX11AccelerationPolicy, MFXVideoCORE_GetHandle error: " +
+                               mfxstatus_to_string(sts));
+    }
+
+    //MFXVideoCORE_SetFrameAllocator(session, mfxFrameAllocator instance)
+    GAPI_LOG_INFO(nullptr, "VPLDX11AccelerationPolicy initialized, session: " << session);
+}
+
+void VPLDX11AccelerationPolicy::deinit(session_t session) {
+    (void)session;
+    GAPI_LOG_INFO(nullptr, "deinitialize session: " << session);
+}
+
+VPLDX11AccelerationPolicy::pool_key_t
+VPLDX11AccelerationPolicy::create_surface_pool(size_t pool_size, size_t surface_size_bytes,
+                                               surface_ptr_ctr_t creator) {
+    GAPI_LOG_DEBUG(nullptr, "pool size: " << pool_size << ", surface size bytes: " << surface_size_bytes);
+
+#ifdef CPU_ACCEL_ADAPTER
+    return adapter->create_surface_pool(pool_size, surface_size_bytes, creator);
+#endif
+    (void)pool_size;
+    (void)surface_size_bytes;
+    (void)creator;
+    throw std::runtime_error(std::string(__FUNCTION__) + " is not implemented");
+}
+
+VPLDX11AccelerationPolicy::surface_weak_ptr_t VPLDX11AccelerationPolicy::get_free_surface(pool_key_t key) const
+{
+#ifdef CPU_ACCEL_ADAPTER
+    return adapter->get_free_surface(key);
+#endif
+    (void)key;
+    throw std::runtime_error(std::string(__FUNCTION__) + " is not implemented");
+}
+
+size_t VPLDX11AccelerationPolicy::get_free_surface_count(pool_key_t key) const {
+    #ifdef CPU_ACCEL_ADAPTER
+    return adapter->get_free_surface_count(key);
+#endif
+    (void)key;
+    throw std::runtime_error(std::string(__FUNCTION__) + " is not implemented");
+}
+
+size_t VPLDX11AccelerationPolicy::get_surface_count(pool_key_t key) const {
+#ifdef CPU_ACCEL_ADAPTER
+    return adapter->get_surface_count(key);
+#endif
+    (void)key;
+    throw std::runtime_error(std::string(__FUNCTION__) + " is not implemented");
```

Comment:
Still, I don't think it's a good idea to use `__FUNCTION__`. It might not be supported in some compilers